src/nti/schema/schema.py

```diff
@@ -38,20 +38,48 @@ def schemadict(spec):
     The schema part (fields) of interface specification *spec* as map
     from name to field.
 
+    The return value should be treated as immutable.
+
     *spec* can be:
 
     - A single Interface;
     - An object implementing ``zope.interfaces.interfaces.ISpecification``,
       such as that returned by ``providedBy(instance)`` or ``implementedBy(factory)``
       (an Interface is a special case of this).
     - A list, tuple, or iterable of Interfaces.
+
+    In the first two cases, the results will be cached using the
+    usual interface caching rules. That means that changes to interface bases,
+    or changes to what an object or class provides or implements, will be properly
+    detected. However, if you simply assign a new field to an existing interface,
+    it may not be detected (things like ``Interface.get()`` also fail in that case)
+    unless you call ``Interface.changed()``.
+
+    .. versionchanged:: 1.15.0
+       Added caching and re-implemented the schemadict algorithm for speed.
     """
-    # ``zope.schema.getFields`` and ``getFieldsInOrder`` deal with a single
-    # interface, only. So in the past, we handled the latter two cases
-    # (which are the most common cases) by constructing a new interface
-    # at runtime using the *spec* as its bases. But that's *really* slow, especially
-    # if the hierarchy is complex. We can do a lot better if we pay attention to what
-    # we're given.
+    try:
+        cache_in = spec._v_attrs # pylint:disable=protected-access
```

Comment:
Even with the comment in the except block It wasn't initially clear to me that we are checking for our first two possibilities, Interface or ISpecification, by the precense of v_attrs. I figured out what was going on here once I realized we require zope.interface > 5.0. Might just have been me.

Reply:
Yeah. `_v_attrs` is an implementation detail so checking `IInterface.providedBy` doesn't actually mean anything.

I should probably see if I can get zope.interface to allow caching with a public API...